components/style_traits/values.rs

```diff
@@ -125,89 +172,15 @@ where
     where
         T: ToCss,
     {
-        if self.writer.has_written {
-            item.to_css(&mut PrefixedWriter::new(&mut self.writer, self.separator))
-        } else {
-            item.to_css(&mut self.writer)
-        }
-    }
-}
-
-struct TrackedWriter<W> {
-    writer: W,
-    has_written: bool,
-}
-
-impl<W> TrackedWriter<W>
-where
-    W: Write,
-{
-    #[inline]
-    fn new(writer: W) -> Self {
-        TrackedWriter {
-            writer: writer,
-            has_written: false,
+        let has_prefix = self.inner.prefix.is_some();
+        if !has_prefix {
+            self.inner.prefix = Some(self.separator);
         }
-    }
-}
-
-impl<W> Write for TrackedWriter<W>
-where
-    W: Write,
-{
-    #[inline]
-    fn write_str(&mut self, s: &str) -> fmt::Result {
-        if !s.is_empty() {
-            self.has_written = true;
+        item.to_css(&mut self.inner)?;
+        if !has_prefix {
+            self.inner.prefix = None;
```

Comment:
I think it'd be nice, more as a self-documenting assert than as a sanity-check, that in this case `self.inner.prefix` is either `None` or `self.separator` before the assignment (since even though it can change down the line, it should always be mutated by another `SequenceWriter`, and thus resetted too.